staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go

```diff
@@ -139,6 +139,31 @@ func NewAvailableConditionController(
 	return c
 }
 
+func setConditionAndUpdateStatus(client apiregistrationclient.APIServicesGetter, apiService *apiregistration.APIService, newCondition apiregistration.APIServiceCondition) error {
+	orig := apiService.DeepCopy()
+	apiregistration.SetAPIServiceCondition(apiService, newCondition)
+	if !reflect.DeepEqual(orig, apiService) {
+		_, err := client.APIServices().UpdateStatus(apiService)
+		if err != nil {
+			return err
+		}
+
+		// update metrics
+		wasTrue := apiregistration.IsAPIServiceConditionTrue(orig, newCondition.Type)
+		switch newCondition.Type {
+		case apiregistration.Available:
+			if wasTrue && newCondition.Status == apiregistration.ConditionFalse {
+				unavailableCounter.WithLabelValues(apiService.Name, newCondition.Reason).Inc()
+				unavailableGauge.WithLabelValues(apiService.Name).Inc()
```

Comment:
Can we directly set this instead?

Reply:
I have a related question: what happens in HA? Multiple instances fight for their (availability) view of the world. How is that handled with a gauge usually. @brancz do you have a suggestion?

Reply:
We need some kind of apiserver identity here and put it into a label

Reply:
re HA: If the API server endpoints are being scraped directly, instance labels can be attached via the Prometheus configuration to each API server target using https://prometheus.io/docs/prometheus/latest/configuration/configuration/#static_config.

If those API servers are sitting behind LBs and are opaque, then things get complicated. If those gauge values are different at different scrapes, those values would be flapping.

Reply:
Nothing should ever be scraped through a load balancer, that would not be whitebox monitoring. And yes what @s-urbaniak said is correct, at ingestion time Prometheus attaches the apiserver's target's labels onto the time-series produced by this metric.

Reply:
@sttts Hi, why do multiple instances have different (availability) view of the world? Could you please help explain it?

Reply:
Because it's a distributed system: different instances are (probably) on different computers with different network connectivity, and they definitely scrape at different times.

Reply:
@lavalamp  So each instance would run below func on `sync()`.  Do you mean in different instances, the parameters `originalAPIService` and `newAPIService` might have different values?
```
func updateAPIServiceStatus(client apiregistrationclient.APIServicesGetter, originalAPIService, newAPIService *apiregistrationv1.APIService) (*apiregistrationv1.APIService, error) {
```